core/arch/arm/plat-stm32mp1/drivers/stm32mp1_rcc.c

```diff
@@ -1,15 +1,64 @@
 // SPDX-License-Identifier: BSD-3-Clause
 /*
- * Copyright (c) 2018, STMicroelectronics
+ * Copyright (c) 2018-2019, STMicroelectronics
  */
 
 #include <drivers/stm32mp1_rcc.h>
+#include <io.h>
+#include <kernel/delay.h>
+#include <kernel/panic.h>
 #include <mm/core_memprot.h>
 #include <platform_config.h>
+#include <stm32_util.h>
+
+#define RESET_ID_MASK		GENMASK_32(31, 5)
+#define RESET_ID_SHIFT		5
+#define RESET_BIT_POS_MASK	GENMASK_32(4, 0)
+
+#define RST_CLR_OFFSET		4U
+#define RESET_TIMEOUT_US	1000
 
 uintptr_t stm32_rcc_base(void)
 {
 	static struct io_pa_va base = { .pa = RCC_BASE };
 
 	return io_pa_or_va(&base);
 }
+
+static size_t reset_id2reg_offset(unsigned int id)
+{
+	return ((id & RESET_ID_MASK) >> RESET_ID_SHIFT) * sizeof(uint32_t);
+}
+
+static uint8_t reset_id2reg_bit_pos(unsigned int reset_id)
+{
+	return (uint8_t)(reset_id & RESET_BIT_POS_MASK);
+}
+
+void stm32_reset_assert(unsigned int id)
+{
+	size_t offset = reset_id2reg_offset(id);
+	uint32_t bitmsk = BIT(reset_id2reg_bit_pos(id));
+	uint64_t timeout_ref = timeout_init_us(RESET_TIMEOUT_US);
+	uintptr_t rcc_base = stm32_rcc_base();
+
+	write32(bitmsk, rcc_base + offset);
+
+	while (!(read32(rcc_base + offset) & bitmsk))
+		if (timeout_elapsed(timeout_ref))
+			panic();
+}
+
+void stm32_reset_deassert(unsigned int id)
+{
+	size_t offset = reset_id2reg_offset(id) + RST_CLR_OFFSET;
+	uint32_t bitmsk = BIT(reset_id2reg_bit_pos(id));
+	uint64_t timeout_ref = timeout_init_us(RESET_TIMEOUT_US);
+	uintptr_t rcc_base = stm32_rcc_base();
```

Comment:
Why `uintptr_t` instead of `vaddr_t` which we usually use for this purpose?

Reply:
I used `uintptr_t` and almost everywhere for such IO registers address. Such function may be called with mmu on or off. I though `uintptr_t` was a clean universal type these  32bit machine drivers.

Reply:
Almost everywhere in the code from ST that is. ;-)
No worries, it's just a bit confusing when we're mixing the types.
The address is acquired from `io_pa_or_va()` which returns a `vaddr_t` and the `write32()` takes a `vaddr_t` as argument.

Reply:
Yes, :))) I meant everywhere in the drivers I've recently upstreamed.
I though OP-TEE core thought about deprecating `vaddr_t` and `paddr_t` in favor to `uintptr_t`. Maybe an old discussion we had I misunderstood.

Reply:
I favor `vaddr_t` and `paddr_t` over `uintptr_t`.
With `CFG_CORE_LARGE_PHYS_ADDR=y` `paddr_t` can be larger than `uintptr_t`.

Reply:
`io_pa_or_va()` returns either a pa or a va but always on a `vaddr_t`. Should return a `paddr_t` then. 

Maybe I should change io_setbits32() and firends (from [io.h](https://github.com/OP-TEE/optee_os/blob/3.4.0/core/include/io.h#L98)) to use `vaddr_t` (or `paddr_t`?) for target address argument.

Reply:
No, when there's no difference between a virtual and physical address then it should be a `vaddr_t`. It's used as a virtual address and has the same effective range as a virtual address.